kmod/core/core.c

```diff
@@ -583,29 +585,57 @@ static int kpatch_kallsyms_callback(void *data, const char *name,
 {
 	struct kpatch_kallsyms_args *args = data;
 
-	if (args->addr == addr && !strcmp(args->name, name))
+	if (strcmp(args->name, name))
+		return 0;
+
+	args->addr = addr;
+	args->count++;
+
+	/*
+	 * Finish the search when the symbol is found for the desired position
+	 * or the position is not defined for a non-unique symbol.
+	 */
+	if ((args->pos && (args->count == args->pos)) ||
+	    (!args->pos && (args->count > 1))) {
 		return 1;
+	}
 
 	return 0;
 }
 
-static int kpatch_verify_symbol_match(const char *name, unsigned long addr)
+static int kpatch_verify_symbol_match(const char *name,
```

Comment:
I agree, we can combine the functionality of `kpatch_verify_symbol_match` with `kpatch_find_module_symbol` into the kpatch-analog version of `klp_find_object_symbol()`. Plus, `kpatch_verify_symbol_match` in the context of kaslr doesn't make sense anymore, since the original function's purpose was to verify if the recorded address (at compile time) in dynrela->src matches the one we find with kallsyms.

In livepatch, `klp_find_object_symbol()` interprets a NULL `objname` argument to mean vmlinux. If the new `kpatch_find_object_symbol()` function is implemented the same way,  we can also get rid of all the `if (vmlinux)` checks and probably keep only the `if (dynrela->external)` checks (though it would be great if we could also get rid of externals eventually).